File: src/bindings/java/in3/eth1/Transaction.java
/*******************************************************************************
 * This file is part of the Incubed project.
 * Sources: https://github.com/slockit/in3-c
 *
 * Copyright (C) 2019 Blockchains, LLC
 * 
 * This program is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free Software 
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *  
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY 
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A 
 * PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.
 * [Permissions of this strong copyleft license are conditioned on making available 
 * complete source code of licensed works and modifications, which include larger 
 * works using a licensed work, under the same license. Copyright and license notices 
 * must be preserved. Contributors provide an express grant of patent rights.]
 * You should have received a copy of the GNU Affero General Public License along 
 * with this program. If not, see <https://www.gnu.org/licenses/>.
 * 
 * If you cannot meet the requirements of AGPL, 
 * you should contact us to inquire about a commercial license.
 *******************************************************************************/

package in3.eth1;

import in3.*;
import java.math.*;

/**
 * represents a Transaction in ethereum.
 * 
 */

public class Transaction {

    private JSON data;

    private Transaction(JSON data) {
        this.data = data;
    }

    protected static Transaction asTransaction(Object o) {
        return o == null ? null : new Transaction((JSON) o);
    }

    /**
     * the blockhash of the block containing this transaction.
     */
    public String getBlockHash() {
        return data.getString("blockHash");
    }

    /**
     * the block number of the block containing this transaction.
     */
    public long getBlockNumber() {
        return data.getLong("blockNumber");
    }

    /**
     * the chainId of this transaction.
     */
    public String getChainId() {
        return data.getString("chainId");
    }

    /**
     * the address of the deployed contract (if successfull)
     */
    public String getCreatedContractAddress() {
        return data.getString("creates");
    }

    /**
     * the address of the sender.
     */
    public String getFrom() {
        return data.getString("from");
    }

    /**
     * the Transaction hash.
     */
    public String getHash() {
        return data.getString("hash");
    }

    /**
     * the Transaction data or input data.
     */
    public String getData() {
        return data.getString("input");
    }

    /**
     * the nonce used in the transaction.
     */
    public long getNonce() {
        return data.getLong("nonce");
    }

    /**
     * the public key of the sender.
     */
    public String getPublicKey() {
        return data.getString("publicKey");
    }

    /**
     * the value send in wei.
     */
    public BigInteger getValue() {
        return data.getBigInteger("value");
    }

    /**
     * the raw transaction as rlp encoded data.
     */
    public String getRaw() {
        return data.getString("raw");
    }

    /**
     * the address of the receipient or contract.
     */
    public String getTo() {
        return data.getString("to");
    }

    /**
     * the signature of the sender - a array of the [ r, s, v]
     */
    public String[] getSignature() {
        return new String[] { data.getString("r"), data.getString("s"), data.getString("v") };
    }

    /**
     * the gas price provided by the sender.
     */
    public long getGasPrice() {
        return data.getLong("gasPrice");
    }

    /**
     * the gas provided by the sender.
     */
    public long getGas() {
        return data.getLong("gas");
    }
}

File: dotnet/In3/Eth1/Transaction.cs
using System.Numerics;
using System.Text.Json.Serialization;
using In3.Utils;

namespace In3.Eth1
{
    /// <summary>
    /// Class representing a transaction that was accepted by the Ethereum chain.
    /// </summary>
    public class Transaction
    {
        /// <summary>
        /// Hash of the block that this transaction belongs to.
        /// </summary>
        [JsonPropertyName("blockHash")] public string BlockHash { get; set; }

        /// <summary>
        /// Number of the block that this transaction belongs to.
        /// </summary>
        [JsonPropertyName("blockNumber"), JsonConverter(typeof(BigIntegerFromHexConverter))] public BigInteger BlockNumber { get; set; }

        /// <summary>
        /// Chain id that this transaction belongs to.
        /// </summary>
        // TODO: swap with a proper chain enum.
        [JsonPropertyName("chainId")] public string ChainId { get; set; }

        /// <summary>
        /// Address of the deployed contract (if successfull).
        /// </summary>
        [JsonPropertyName("creates")] public string Creates { get; set; }


        /// <summary>
        /// Address whose private key signed this transaction with.
        /// </summary>
        [JsonPropertyName("from")] public string From { get; set; }

        /// <summary>
        /// Gas for the transaction.
        /// </summary>
        [JsonPropertyName("gas"), JsonConverter(typeof(LongFromHexConverter))] public long Gas { get; set; }


        /// <summary>
        /// Gas price (in wei) for each unit of gas.
        /// </summary>
        [JsonPropertyName("gasPrice"), JsonConverter(typeof(LongFromHexConverter))] public long GasPrice { get; set; }

        /// <summary>
        /// Transaction hash.
        /// </summary>
        [JsonPropertyName("hash")] public string Hash { get; set; }

        /// <summary>
        /// Transaction data.
        /// </summary>
        [JsonPropertyName("input")] public string Input { get; set; }

        /// <summary>
        /// Nonce for this transaction.
        /// </summary>
        [JsonPropertyName("nonce"), JsonConverter(typeof(LongFromHexConverter))] public long Nonce { get; set; }

        /// <summary>
        /// Public key.
        /// </summary>
        [JsonPropertyName("publicKey")] public string PublicKey { get; set; }

        /// <summary>
        /// Part of the transaction signature.
        /// </summary>
        [JsonPropertyName("r")] public string R { get; set; }
        /// <summary>
        /// Transaction as rlp encoded data.
        /// </summary>
        [JsonPropertyName("raw")] public string Raw { get; set; }

        /// <summary>
        /// Part of the transaction signature.
        /// </summary>
        [JsonPropertyName("s")] public string S { get; set; }

        /// <summary>
        /// Part of the transaction signature. V is parity set by v = 27 + (r % 2).
        /// </summary>
        /// <value>Either <see langword="0" /> or <see langword="1" />.</value>
        [JsonPropertyName("standardV")] public string StandardV { get; set; }

        /// <summary>
        /// To address of the transaction.
        /// </summary>
        [JsonPropertyName("to")] public string To { get; set; }

        /// <summary>
        /// Transaction index.
        /// </summary>
        // TODO: Check if this doesnt make more sense as a long.
        [JsonPropertyName("transactionIndex")] public string TransactionIndex { get; set; }

        /// <summary>
        /// The <see cref="StandardV" /> plus the chain.
        /// </summary>
        [JsonPropertyName("v")] public string V { get; set; }

        /// <summary>
        /// Value of the transaction.
        /// </summary>
        [JsonPropertyName("value"), JsonConverter(typeof(BigIntegerFromHexConverter))] public BigInteger Value { get; set; }
    }
}